File: python/minimum_window_substring.py
class Solution:
    def minWindow(self, s: str, t: str) -> str:
        if t == "":
            return ""

        result: list[int] = []
        resultLen: float = float("inf")

        have_map: dict[str, int] = {}
        need_map: dict[str, int] = {}
        for c in t:
            need_map[c] = 1 + need_map.get(c, 0)
        need = len(need_map)
        have = 0

        l = 0
        for r in range(len(s)):
            r_c = s[r]
            have_map[r_c] = 1 + have_map.get(r_c, 0)
            if r_c in need_map and have_map[r_c] == need_map[r_c]:
                have += 1

            while have == need:
                if r + 1 - l < resultLen:
                    result = [l, r]
                    resultLen = r + 1 - l

                l_c = s[l]
                have_map[l_c] -= 1
                if l_c in need_map and have_map[l_c] + 1 == need_map[l_c]:
                    have -= 1
                l += 1

        return s[result[0] : result[1] + 1] if resultLen != float("inf") else ""


def test_permutation_in_string_case_1():
    # arrange
    s = "ADOBECODEBANC"
    t = "ABC"
    expected = "BANC"

    # act
    solution = Solution()
    actual = solution.minWindow(s, t)

    # assert
    assert actual == expected


def test_permutation_in_string_case_2():
    # arrange
    s = "a"
    t = "a"
    expected = "a"

    # act
    solution = Solution()
    actual = solution.minWindow(s, t)

    # assert
    assert actual == expected


def test_permutation_in_string_case_3():
    # arrange
    s = "a"
    t = "aa"
    expected = ""

    # act
    solution = Solution()
    actual = solution.minWindow(s, t)

    # assert
    assert actual == expected


def test_permutation_in_string_case_4():
    # arrange
    s = "aab"
    t = "aab"
    expected = "aab"

    # act
    solution = Solution()
    actual = solution.minWindow(s, t)

    # assert
    assert actual == expected

File: rust/src/code/minimum_window_substring.rs
use std::collections::HashMap;

pub struct Solution {}

impl Solution {
    pub fn min_window(s: String, t: String) -> String {
        let mut min_intervals: (usize, usize) = (0, 0);
        let mut interval_len: usize = usize::MAX;
        let mut have = 0;
        let mut have_map: HashMap<char, i32> = HashMap::new();
        let mut need_map: HashMap<char, i32> = HashMap::new();
        for c in t.chars() {
            need_map.entry(c).and_modify(|v| *v += 1).or_insert(1);
        }
        let need = need_map.keys().len();
        let s_chars: Vec<char> = s.chars().collect();

        let mut l = 0;
        for r in 0..s_chars.len() {
            let c_r = s_chars[r];
            have_map
                .entry(c_r)
                .and_modify(|v| {
                    *v += 1;
                })
                .or_insert(1);

            match need_map.get(&c_r) {
                Some(&n_v) => {
                    if let Some(&h_v) = have_map.get(&c_r) {
                        if n_v == h_v {
                            have += 1;
                        }
                    }
                }
                None => (),
            }

            while have == need {
                Self::update_min_interval(&mut min_intervals, &mut interval_len, l, r);

                let c_l = s_chars[l];

                have_map.entry(c_l).and_modify(|v| *v -= 1);
                match need_map.get(&c_l) {
                    Some(&v_n) => {
                        if let Some(&v_h) = have_map.get(&c_l) {
                            if v_n == v_h + 1 {
                                have -= 1;
                            }
                        }
                    }
                    None => (),
                }
                l += 1;
            }
        }

        if interval_len == usize::MAX {
            return String::new();
        } else {
            return s[min_intervals.0..=min_intervals.1].to_string();
        }
    }

    fn update_min_interval(
        min_interval: &mut (usize, usize),
        min_interval_len: &mut usize,
        l: usize,
        r: usize,
    ) {
        if (r + 1 - l) < *min_interval_len {
            *min_interval = (l, r);
            *min_interval_len = r + 1 - l;
        }
    }
}

#[cfg(test)]
mod tests {
    use super::Solution;

    #[test]
    fn minimum_window_substring_case_1() {
        // arrange
        let s = String::from("ADOBECODEBANC");
        let t = String::from("ABC");
        let expected = String::from("BANC");

        // act
        let actual = Solution::min_window(s, t);

        // assert
        assert_eq!(expected, actual);
    }

    #[test]
    fn minimum_window_substring_case_2() {
        // arrange
        let s = String::from("a");
        let t = String::from("a");
        let expected = String::from("a");

        // act
        let actual = Solution::min_window(s, t);

        // assert
        assert_eq!(expected, actual);
    }

    #[test]
    fn minimum_window_substring_case_3() {
        // arrange
        let s = String::from("a");
        let t = String::from("aa");
        let expected = String::from("");

        // act
        let actual = Solution::min_window(s, t);

        // assert
        assert_eq!(expected, actual);
    }
}
